quil-rs/src/program/scheduling/schedule.rs

```diff
@@ -0,0 +1,367 @@
+    duration: Time,
+}
+
+impl<Time> Schedule<Time> {
+    pub fn duration(&self) -> &Time {
+        &self.duration
+    }
+
+    pub fn items(&self) -> &[ComputedScheduleItem<Time>] {
+        self.items.as_ref()
+    }
+}
+
+impl<Time: Zero> Default for Schedule<Time> {
+    fn default() -> Self {
+        Self {
+            items: Default::default(),
+            duration: Time::zero(),
+        }
+    }
+}
+
+pub type FixedSchedule = Schedule<Seconds>;
+
+#[derive(Debug)]
+pub struct ComputedScheduleItem<Time> {
+    pub time_span: TimeSpan<Time>,
+    pub instruction_index: usize,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ComputedScheduleError {
+    #[error("unknown duration for instruction {}", instruction.to_quil_or_debug())]
+    UnknownDuration { instruction: Instruction },
+
+    #[error("internal error: invalid dependency graph")]
+    InvalidDependencyGraph,
+}
+
+pub type ComputedScheduleResult<T> = Result<T, ComputedScheduleError>;
+
+/// Represents a span of time, for some unit of time
+#[derive(Debug)]
+pub struct TimeSpan<Time> {
+    /// The inclusive start time of the described item
+    start_time: Time,
+
+    /// The described item's continuous duration
+    duration: Time,
+}
+
+impl<Time> TimeSpan<Time> {
+    pub fn start_time(&self) -> &Time {
+        &self.start_time
+    }
+
+    pub fn duration(&self) -> &Time {
+        &self.duration
+    }
+}
+
+impl<'p> ScheduledBasicBlock<'p> {
+    // todo create more restricted subtype of instruction for ScheduledBasicBlock so we don't have
+    // to handle unreachable instructions here
+    /// Return the duration of a scheduled Quil instruction:
+    ///
+    /// * For PULSE and CAPTURE, this is the duration of the waveform at the frame's sample rate
+    /// * For DELAY and RAW-CAPTURE, it's the named duration
+    /// * For supporting instructions like SET-*, SHIFT-*, and FENCE, it's 0
+    ///
+    /// Return `None` for other instructions.
+    fn get_fixed_instruction_duration(
+        program: &Program,
+        instruction: &Instruction,
+    ) -> Option<Seconds> {
+        match instruction {
+            Instruction::Capture(Capture { waveform, .. })
+            | Instruction::Pulse(Pulse { waveform, .. }) => {
+                Self::get_fixed_waveform_duration(program, instruction, waveform)
+            }
+            Instruction::Delay(Delay { duration, .. })
+            | Instruction::RawCapture(RawCapture { duration, .. }) => {
+                duration.to_real().ok().map(Seconds)
+            }
+            Instruction::Fence(_)
+            | Instruction::SetFrequency(_)
+            | Instruction::SetPhase(_)
+            | Instruction::SetScale(_)
+            | Instruction::ShiftFrequency(_)
+            | Instruction::ShiftPhase(_)
+            | Instruction::SwapPhases(_) => Some(Seconds(0.0)),
+            // Todo should this tolerate classical instructions? how about reset?
```

Comment:
This may depend on the way we use `subenum` if at all, but todo, either (a) confirm that classical instructions are skipped by the graph visitor and never called here or (b) report a duration of 0 for them.

In any case, there should be test cases covering this.